### ISSUE TITLE
sequins: changes to flow modifiers nest, rather than dynamically modify

### ISSUE THREAD

**Issue by dndrks** [monome/crow#435]
while working on norns `sequins` docs, I realized that the pattern established by `every` doesn't apply to the flow modifiers (both on crow and in the norns library). the gesture i was testing was whether an inner-nested sequins could have its flow modifiers changed on-the-fly / without recompiling the outer sequins.

example:

```lua
inner = s{333,444,555}:count(5)
outer = s{1,2,4,6,inner}
```

executing `outer()` returns:
```lua
1
2
4
6
333 -- rep 1
444 -- rep 2
555 -- rep 3
333 -- rep 4
444 -- rep 5
1 -- loop point
```

if i try to dynamically change inner's `count` (`inner:count(2)`), the results from `outer()` are:
```lua
1
2
4
6
555
333
444
555
333
444
555
333
444
555
1
```

seems like it's `count`ing the previous `count`, which _is_ super cool, but unexpected.
also, `inner:reset()` doesn't reset the `count`.

lmk if i can help with any other context / examples! thanks t <333

**Comment by trentgill** [comment]
just some notes for reference in case someone wants to pick this up.

at present, when you use flow-modifiers, they are stacked-up in a nested table in the sequins object. thus execution is sequential through the list -- importantly this means that `:count(3):every(2)` is different to `:every(2):count(3)`. the former gives pattern {0,{1,1,1}} and the latter: {{0,1,0},{1,0,1}}, where inner-braces capture execution of the parent.

when you apply a `:count` to an object that already has a count it just appends it: `:count():count()` hence the behaviour noted above.

the above is in contrast to the intrinsic `:step` method which alters a fixed element of the sequins itself. i agree that this is a confusion distinction because the two methods look from the outside to be the same kind of operation.

-----------------------------

i see 2 options:

1) make `count`, `every` & `times` intrinsic properties of a sequins, rather than arbitrary methods chained onto the object. this would require a flag / table to note the order of configuration. (what happens when applying `:count` to a sequins object where the count was defined before `every` but now it's attached after?)

core downside i see is that it concretizes the implementation and makes extending sequins in the future a little more involved (you would have to change the structure / content of a sequins, rather than just add a new method).

2) alternative approach is to make it easier to hotswap an entire sequins, maintaining internal counters, such that the idiomatic approach is to simply redefine the whole sequins (with eg. a new `count` value), rather than apply it as a piecemeal method.

perhaps someone has a better idea?!

**Comment by trentgill** [comment]
been thinking about this and realizing that it would be preferable for there to be a fixed execution / realization of the flow modifers.

this means it shouldn't make a difference if you `:count(2):every(3)` vs `:every(3):count(2)`. instead we have a fixed order of application:

1. check if `every` should return a value this iteration ->
2. check if `count` should steal the next value too (overrides the next `every`)
3. decrement any `times` counter, setting a 'dead' flag if it reaches 0

this means we can remove the recursive modifier stack, and store any modifiers as keys in the sequins table along, where value is the argument. additionally we'll track an execution state for each modifier.

the solution is thus less extensible (one has to change the structure of sequins itself), but from seeing it in use over the past few months, it doesn't seem that there are obvious holes in the implementation.

///

this data-driven representation of flow modifiers makes it easy & predictable to modify the values, as well as being easier to predict what value a sequins will produce (rather than arbitrarily stacking up modifiers & avoiding order-of-operations concerns).

we will use a similar approach to extending sequins with `transforms`, where it's just a key in the table with some behaviour attached.

**Comment by trentgill** [comment]
these changes are merged in #446

see that PR for more discussion on the motivations & particular solution.
